sdk/rust/oak/src/http/client.rs

```diff
@@ -0,0 +1,84 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{
+    http::{util::Pipe, Invocation},
+    OakError,
+};
+use http::{Request, Response};
+use oak_abi::label::{confidentiality_label, tls_endpoint_tag, Label};
+use oak_io::Sender;
+use oak_services::proto::oak::encap::HttpRequest;
+use std::convert::TryInto;
+
+/// Initializes an HTTP client pseudo-Node serving requests that have the given authority in their
+/// URIs. If the authority is empty, the node is public and can handle any incoming request.
+/// Otherwise, the node is created with the `authority` as its confidentially tag. In this case, the
+/// client can only serve HTTPS requests to the services indicated by the `authority`.
+///
+/// Returns an [`HttpClient`].
+pub fn init(authority: &str) -> Result<HttpClient, OakError> {
+    let config = crate::node_config::http_client(authority);
+    let label = if authority.is_empty() {
+        Label::public_untrusted()
+    } else {
+        confidentiality_label(tls_endpoint_tag(authority))
+    };
+    let sender = crate::io::node_create::<Invocation>("http_client", &label, &config)?;
+    Ok(HttpClient {
+        sender,
+        authority: authority.to_string(),
+    })
+}
+
+pub struct HttpClient {
+    sender: Sender<Invocation>,
+    authority: String,
+}
+
+impl HttpClient {
+    /// Sends the given request via the HTTP client pseudo-node. The `response_label` should match
+    /// the label of the Oak Node.
```

Comment:
There seems to be no enforcement of this "should", which means that a node may just set that to public, send a request to some endpoint that reflects back the request as-is (or returns data which is as sensitive as the request, anyways), and this would have the effect of declassifying any data labelled with this label, right?

It seems to me that the response label should just always be set to be the same of the request label, but I haven't thought about it enough yet.

Reply:
I think that makes sense actually. Imagine there is some data labelled with TLS:google.com (i.e. intended to be sent to google.com). If we allow the caller to set the label of the response, it can just set it to public, send the data to google.com, e.g. to the search page, and then the result page (which of course also contains the query itself) would be public, which is effectively a declassification! So whatever we end up doing, this kind of attack must be prevented. I think the simplest thing is to set the label of the response to be the same as the request, which intuitively means that the data is not declassified in the round trip over HTTP / gRPC. WDYT?

Reply:
I guess because you have removed the downgrades from Pipe now this is actually implicitly enforced, so actually setting this label to public when the HTTP client node is non-public would result in the response being lost. Is that correct?

Reply:
(either way, this seems worth a test case, to show that an HTTP round trip may not be used to declassify data)

Reply:
For instance in the /test_google endpoint, if it used an actual non-public label, I would expect that the response label cannot be set to public (though this cannot be tested currently because the label is public).

Reply:
> I think the simplest thing is to set the label of the response to be the same as the request, which intuitively means that the data is not declassified in the round trip over HTTP / gRPC. WDYT?

In that case, the SDK (in `send_request`) should set the label on both the request and response channels based on the `authority` of the HTTP client node. The Oak node must be created with the same label too. So, I think in the general case it means that we need something like the router/virtual node pattern here; to create a separate Oak Node for each authority. 

> so actually setting this label to public when the HTTP client node is non-public would result in the response being lost. Is that correct?

Yes, I think if the HTTP client node is non-public, then it won't be able to write into a public channel. I noticed that we don't have a test for this. It should be added to the runtime tests in a separate PR. 

> For instance in the /test_google endpoint, if it used an actual non-public label, I would expect that the response label cannot be set to public (though this cannot be tested currently because the label is public).

Currently, the runner runs the examples with the `root_tls_certificate` set to `./examples/certs/local/ca.pem`. So, we cannot test this scenario (a non-public HTTP client with a non-empty authority) in the `http_server` example.

Reply:
> > I think the simplest thing is to set the label of the response to be the same as the request, which intuitively means that the data is not declassified in the round trip over HTTP / gRPC. WDYT?
> 
> In that case, the SDK (in `send_request`) should set the label on both the request and response channels based on the `authority` of the HTTP client node. The Oak node must be created with the same label too. So, I think in the general case it means that we need something like the router/virtual node pattern here; to create a separate Oak Node for each authority.

I think just setting the label to be the same between request and responses fixes all these issues. If it's public, then there is no issue, they can both be public. If it has a TLS label, then the request will have that label, but also the response, which intuitively is as confidential as the request (in general). If the application handles multiple TLS labels, it needs to create multiple nodes, one per authority. That seems sensible, regardless of what we do about the response label.

> > so actually setting this label to public when the HTTP client node is non-public would result in the response being lost. Is that correct?
> 
> Yes, I think if the HTTP client node is non-public, then it won't be able to write into a public channel. I noticed that we don't have a test for this. It should be added to the runtime tests in a separate PR.

SG

> > For instance in the /test_google endpoint, if it used an actual non-public label, I would expect that the response label cannot be set to public (though this cannot be tested currently because the label is public).
> 
> Currently, the runner runs the examples with the `root_tls_certificate` set to `./examples/certs/local/ca.pem`. So, we cannot test this scenario (a non-public HTTP client with a non-empty authority) in the `http_server` example.

I don't remember the details, isn't that root TLS cert for the HTTP *server* part?

Reply:
Created #1775 that updates the HTTP client tests in the runtime. Everything is either public or has the tls endpoint as its label. The Oak node needs to have a privilege too. Would it be able to create channels without downgrading since it is not public? I guess some of those downgrades are needed in the SDK too. 

> I don't remember the details, isn't that root TLS cert for the HTTP server part?

Does not seem like it. We have `grpc_tls_certificate` and `http_tls_certificate` for the server nodes. If `root_tls_certificate` is left empty, then the `oak_loader` uses the default root certificate. But I am not sure what that is.  Is that actually a valid certificate? Perhaps @conradgrobler knows.

Reply:
The "default root certificate" is actually many root certificates. It is generated from from the list of CA root certificates trusted by the Android Open Source project. See https://github.com/project-oak/oak/tree/main/oak_loader/src/certs

Reply:
IIRC that is because we want to use TLS but connecting to servers for testing, e.g. aggregator backend. Therefore, it seems it is not possible to have both regular root CAs and custom CAs, we should just make the test use a custom local back-end over TLS instead of going out to google.com.

Reply:
With HTTPS requests we need a domain name in the uri. So, a local backend won't work. We could perhaps use `oakexamples.dev`. But for that we'll need to have an HTTP server with a self-signed certificate on GCP, similar to what we currently have for the aggregator example with a gRPC server.

Reply:
I think we can self-sign a cert for localhost or even 127.0.0.1, isn't that the point of providing a custom CA for testing?

Reply:
There is a problem with connecting to 127.0.0.1 over TLS, as the web pki crate that we are using does not support validating certificates for an IP. It should still be fine to use localhost for the connections.

Reply:
BTW maybe we should consider making the logic add the root CA specified on the command line to the ones that we load from the file, rather than replace all of them. It may make this kind of test easier.

Reply:
In #1779 I have replaced the request to google.com with a request to localhost, and the HTTP client pseudo-node now has a non-public label. But it is not possible to make this example work without at least one node that has the privilege to declassify responses from the HTTP client node. I think, the only node that makes sense to have that kind of privilege is a cryptographic node that encrypts and declassifies the response. So the changes in #1779 should be postponed until we have that pseudo node. Or are the any other solutions?

Reply:
Not sure I follow the argument, is it that because a crypto node would leak some information then we can use it to bypass IFC rules?

Reply:
(also should this discussion be on #1779 maybe?)